controllers/volsync/vshandler.go

```diff
@@ -1571,3 +1662,376 @@ func ValidateObjectExists(ctx context.Context, c client.Client, obj client.Objec
+	if err != nil {
+		return lrd, nil, err
+	}
+
+	l.V(1).Info(fmt.Sprintf("Local ReplicationDestination Reconcile Complete lrd=%s,lrs=%s", lrd.Name, lrs.Name))
+
+	return lrd, lrs, nil
+}
+
+func (v *VSHandler) reconcileLocalRD(rdSpec ramendrv1alpha1.VolSyncReplicationDestinationSpec,
+	pskSecretName string) (*volsyncv1alpha1.ReplicationDestination, error,
+) {
+	l := v.log.WithValues("rdSpec", rdSpec)
+
+	lrd := &volsyncv1alpha1.ReplicationDestination{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      getLocalReplicationName(rdSpec.ProtectedPVC.Name),
+			Namespace: v.owner.GetNamespace(),
+		},
+	}
+
+	err := v.EnsurePVCforDirectCopy(v.ctx, l, rdSpec)
+	if err != nil {
+		return nil, err
+	}
+
+	op, err := ctrlutil.CreateOrUpdate(v.ctx, v.client, lrd, func() error {
+		if err := ctrl.SetControllerReference(v.owner, lrd, v.client.Scheme()); err != nil {
+			l.Error(err, "unable to set controller reference")
+
+			return err
+		}
+
+		addVRGOwnerLabel(v.owner, lrd)
+		addLabel(lrd, VolSyncDoNotDeleteLabel, VolSyncDoNotDeleteLabelVal)
+
+		pvcAccessModes := []corev1.PersistentVolumeAccessMode{corev1.ReadWriteOnce} // Default value
+		if len(rdSpec.ProtectedPVC.AccessModes) > 0 {
+			pvcAccessModes = rdSpec.ProtectedPVC.AccessModes
+		}
+
+		lrd.Spec.RsyncTLS = &volsyncv1alpha1.ReplicationDestinationRsyncTLSSpec{
+			ServiceType: v.getRsyncServiceType(),
+			KeySecret:   &pskSecretName,
+
+			ReplicationDestinationVolumeOptions: volsyncv1alpha1.ReplicationDestinationVolumeOptions{
+				CopyMethod:       volsyncv1alpha1.CopyMethodDirect,
+				Capacity:         rdSpec.ProtectedPVC.Resources.Requests.Storage(),
+				StorageClassName: rdSpec.ProtectedPVC.StorageClassName,
+				AccessModes:      pvcAccessModes,
+				DestinationPVC:   &rdSpec.ProtectedPVC.Name,
+			},
+		}
+
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	// Now check status - only return an RD if we have an address filled out in the ReplicationDestination Status
+	if lrd.Status == nil || lrd.Status.RsyncTLS == nil || lrd.Status.RsyncTLS.Address == nil {
+		l.V(1).Info("Local ReplicationDestination waiting for Address ...")
+
+		return nil, nil
+	}
+
+	l.V(1).Info("Local ReplicationDestination Reconcile Complete", "op", op)
+
+	return lrd, nil
+}
+
+//nolint:funlen
+func (v *VSHandler) reconcileLocalRS(rd *volsyncv1alpha1.ReplicationDestination,
+	rsSpec *ramendrv1alpha1.VolSyncReplicationSourceSpec,
+	pskSecretName, address string,
+) (*volsyncv1alpha1.ReplicationSource, error,
+) {
+	storageClass, err := v.getStorageClass(rsSpec.ProtectedPVC.StorageClassName)
+	if err != nil {
+		return nil, err
+	}
+
+	volumeSnapshotClassName, err := v.getVolumeSnapshotClassFromPVCStorageClass(storageClass)
+	if err != nil {
+		return nil, err
+	}
+
+	// Fix for CephFS (replication source only) - may need different storageclass and access modes
+	err = v.ModifyRSSpecForCephFS(rsSpec, storageClass)
+	if err != nil {
+		return nil, err
+	}
+
+	pvc, err := v.setupLocalRS(rd)
```

Comment:
I think this part still makes me nervous - if I'm reading it correctly if the latestImage is updated on the RD, then we'll reconcile the localRS again and switch it out with the new latest image - possibly trying to create a new read-only PVC, etc etc.

This means if the previous transfer is not done, we'll be updating the localRS while it's in transfer - and potentially getting into a loop where it doesn't ever complete.

Here's a scenario:
1 - initial replication is setup, PVC is large and the first transfer takes a long time
2 - localRS/RD setup after 1st latestImage becomes available
3- in the meantime the next transfer has already started since perhaps it missed the initial window
4- new latestImage is available
5- the previous localRS/RD sync is not complete, but we replace the latestImage while it's running, leading to issues or the local transfer not completing since we keep reconfiguring the localRS with different read-only pvcs as the source.

This is the hardest problem to solve, and I would think it's safer to never replace while a transfer is taking place

Reply:
That scenario is covered here:
https://github.com/RamenDR/ramen/blob/fa43e3a049b80d8e21812f3846f939bcc5136eea/controllers/volsync/vshandler.go#L1950

Unless I am missing something.

Reply:
sorry, I didn't read it through carefully enough - the cleanupPrevious is exactly what I was looking for and does mean we don't replace the pvc the localRS is using while it's syncing